src/ibek/entity_factory.py

```diff
@@ -0,0 +1,177 @@
+        for value in definition.values:
+            add_arg(value.name, str, value.description, value.value)
+
+        # add the type literal which discriminates between the different Entity classes
+        typ = Literal[full_name]  # type: ignore
+        add_arg("type", typ, definition.description, full_name)
+
+        class_name = full_name.replace(".", "_")
+        entity_cls = create_model(
+            class_name,
+            **args,
+            __validators__=validators,
+            __base__=Entity,
+        )  # type: ignore
+
+        # add a link back to the Definition Instance that generated this Entity Class
+        entity_cls.__definition__ = definition
+
+        # store this Entity class in the factory
+        self._entity_models[full_name] = entity_cls
+
+        return entity_cls
+
+    def _make_entity_models(self, support: Support) -> List[Type[Entity]]:
+        """
+        Create Entity subclasses for all Definition instances in the given
+        Support instance. Returns a list of the Entity subclasses Models.
+        """
+        entity_names = []
+        entity_models = []
+
+        for definition in support.defs:
+            if definition.name in entity_names:
+                # not tested because schema validation will always catch this first
+                raise ValueError(f"Duplicate entity name {definition.name}")
+
+            entity_models.append(self._make_entity_model(definition, support))
+
+            entity_names.append(definition.name)
+        return entity_models
+
+    def process_collections(self, entities: List[Entity]):
+        """
+        Process all the SubEntity collections in a list of Entity instances
+        """
+        all_entities: List[Entity] = []
+
+        def resolve_sub_entities(entity: Entity):
+            # recursive function to scan for SubEntities in an entity
+            definition = entity.__definition__
+
+            # add the parent standard entity
+            all_entities.append(entity)
+
+            # add in SubEntities if any
+            for sub_entity in definition.sub_entities:
+                # find the Entity Class that the SubEntity represents
+                entity_cls = self._entity_models[sub_entity.type]
+                # get the SubEntity arguments
+                sub_args_dict = sub_entity.model_dump()
+                # jinja render any references to parent Args in the SubEntity Args
+                for key, arg in sub_args_dict.items():
+                    sub_args_dict[key] = UTILS.render(entity, arg)
+                # cast the SubEntity to its concrete Entity subclass
+                cast_entity = entity_cls(**sub_args_dict)
+                # recursively scan the SubEntity for more SubEntities
+                resolve_sub_entities(cast_entity)
+
+        for entity in entities:
+            resolve_sub_entities(entity)
+
+        return all_entities
```

Comment:
I think it would be nicer to flatten this nested function. What about this? (Also renaming some things)

```suggestion
    def resolve_sub_entities(self, entities: List[Entity]) -> List[Entity]:
        """
        Recursively resolve SubEntity collections in a list of Entity instances
        """
        resolved_entities: List[Entity] = []
        for parent_entity in entities:
            definition = parent_entity.__definition__

            # add the parent standard entity
            resolved_entities.append(parent_entity)

            # add in SubEntities if any
            for sub_entity in definition.sub_entities:
                # find the Entity Class that the SubEntity represents
                entity_cls = self._entity_models[sub_entity.type]
                # get the SubEntity arguments
                sub_args_dict = sub_entity.model_dump()
                # jinja render any references to parent Args in the SubEntity Args
                for key, arg in sub_args_dict.items():
                    sub_args_dict[key] = UTILS.render(parent_entity, arg)
                # cast the SubEntity to its concrete Entity subclass
                entity = entity_cls(**sub_args_dict)
                # recursively scan the SubEntity for more SubEntities
                resolved_entities.extend(self.resolve_sub_entities([entity]))

        return resolved_entities
```

Reply:
Nice - will use this.